device/common/include/traccc/device/unique_lock.hpp

```diff
@@ -0,0 +1,76 @@
+/**
+ * traccc library, part of the ACTS project (R&D line)
+ *
+ * (c) 2024 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+#include <mutex>
+#include <vecmem/memory/device_atomic_ref.hpp>
+
+#include "traccc/definitions/qualifiers.hpp"
+
+namespace traccc::device {
+template <typename Mutex>
+class unique_lock {
+    public:
+    using mutex_type = Mutex;
+
+    /*
+     * Construct a unique lock without locking.
+     */
+    TRACCC_HOST_DEVICE
+    unique_lock(mutex_type& m, std::defer_lock_t);
+
+    /*
+     * Construct a unique lock, attempting to lock it.
+     */
+    TRACCC_HOST_DEVICE
+    unique_lock(mutex_type& m, std::try_to_lock_t);
+
+    /*
+     * Construct a unique lock which was previously locked.
+     */
+    TRACCC_HOST_DEVICE
+    unique_lock(mutex_type& m, std::adopt_lock_t);
+
+    /*
+     * Destroy a lock, freeing the underlying mutex.
+     */
+    TRACCC_HOST_DEVICE
+    ~unique_lock();
+
+    /*
+     * Lock the lock, blocking until the operation succeeds.
+     */
+    TRACCC_HOST_DEVICE
+    void lock();
+
+    /*
+     * Try to lock the lock without blocking.
+     */
+    TRACCC_HOST_DEVICE
+    bool try_lock();
+
+    /*
+     * Explicitly lock the underlying lock.
+     */
+    TRACCC_HOST_DEVICE
+    void unlock();
+
+    /*
+     * Check if the lock is locked by this object.
+     */
+    TRACCC_HOST_DEVICE
+    bool owns_lock() const;
+
+    private:
+    mutex_type* m_mutex_ptr = nullptr;
```

Comment:
Would a reference not be better? These locks are not meant to be copiable/movable anyway, right?

Reply:
This is a subset of `std::unique_lock`'s semantics, which also support the ability for a lock not to be associated with any mutex at all. If we ever want to extend `traccc::device::unique_lock` to support the same semantics, this would be a good start. `std::unique_lock` is movable.